domain/src/main/java/org/sagebionetworks/research/domain/task/navigation/TreeNavigator.java

```diff
@@ -0,0 +1,336 @@
+           this.progressMarkers = null;
+       } else {
+           ImmutableList.Builder<String> builder = new ImmutableList.Builder<>();
+           builder.addAll(progressMarkers);
+           this.progressMarkers = builder.build();
+       }
+       this.stepsById = buildStepsByID(steps);
+    }
+
+    /**
+     * Constructs and returns an ImmutableMap that maps step identifiers to steps for every step
+     * and every substep of every step recursively, in the given list of steps.
+     * @param steps The list of steps to construct the map from.
+     * @return An ImmutableMap that maps step identifiers to steps for every step and every substep
+     *         of every step recursively, in the given list of steps.
+     */
+    private static ImmutableMap<String, Step> buildStepsByID(List<Step> steps) {
+        ImmutableMap.Builder<String, Step> builder = new ImmutableMap.Builder<>();
+        for (Step step : steps) {
+            addStepToBuilderRecursively(step, builder);
+        }
+
+        return builder.build();
+    }
+
+    /**
+     * Adds the given step and all of its substeps to the given builder.
+     * @param step The step to add (along with all of its substeps) to the given builder.
+     * @param builder The builder to add the given step to.
+     */
+    private static void addStepToBuilderRecursively(Step step, ImmutableMap.Builder<String, Step> builder) {
+        if (step instanceof SectionStep) {
+            SectionStep sectionStep = (SectionStep)step;
+            builder.put(sectionStep.getIdentifier(), sectionStep);
+            for (Step childStep : sectionStep.getSteps()) {
+                addStepToBuilderRecursively(childStep, builder);
+            }
+        } else {
+            builder.put(step.getIdentifier(), step);
+        }
+    }
+
+    @Nullable
+    @Override
+    public Step getStep(@NotNull String identifier) {
+        return this.stepsById.get(identifier);
+    }
+
+    @Nullable
+    @Override
+    public Step getNextStep(@Nullable Step step, @NotNull TaskResult taskResult) {
+        return nextStepHelper(step, this.root, new AtomicBoolean(false));
+    }
+
+    /**
+     * Returns the first leaf step that appears after the given initialStep, in this TreeNavigator.
+     * After is defined as the next leaf in a pre-order traversal of the tree.
+     * @param initialStep The step to find the step after.
+     * @param current The step that is being evaluated by this call to nextStepHelper().
+     * @param hasFoundInitial True if the initialStep has already been encountered by a parent
+     *                        recursive call of this call to nextStepHelper(), false otherwise.
+     * @return The first leaf step that appears after the given initialStep, or null if no such
+     *         step exists.
+     */
+    @Nullable
+    private static Step nextStepHelper(@Nullable Step initialStep, @Nullable Node current,
+                                       AtomicBoolean hasFoundInitial) {
```

Comment:
Why an AtomicBoolean here and not the primitive boolean?

Reply:
The boolean needs to be mutated here because we are doing a pre-order traversal and the next leaf isn't necessarily below the leaf with the initial step in the tree. Making this an AtomicBoolean allows the mutation to still be reflected when we go back up the call stack without having to do anything funky with return values.

Reply:
ah cool, missed that part